eventsourcing/src/main/java/org/axonframework/eventsourcing/eventstore/EventStoreTransaction.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2010-2024. Axon Framework
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.eventsourcing.eventstore;
+
+import jakarta.annotation.Nonnull;
+import org.axonframework.eventhandling.EventMessage;
+import org.axonframework.messaging.MessageStream;
+import org.axonframework.messaging.unitofwork.ProcessingContext;
+
+import java.util.function.Consumer;
+
+/**
+ * Interface describing the actions that can be taken on a transaction to source a model from the
+ * {@link AsyncEventStore} based on the resulting {@link MessageStream}.
+ * <p>
+ * Note that this transaction includes operations for {@link #source(SourcingCondition, ProcessingContext) sourcing} the
+ * model as well as {@link #appendEvent(EventMessage) appending events}.
+ *
+ * @author Allard Buijze
+ * @author Steven van Beelen
+ * @since 5.0.0
+ */
+public interface EventStoreTransaction {
+
+    /**
+     * A {@link org.axonframework.messaging.unitofwork.ProcessingContext.ResourceKey} referring to the position of the
+     * last {@link #appendEvent(EventMessage) appended event}.
+     */
+    ProcessingContext.ResourceKey<Long> APPEND_POSITION_KEY = ProcessingContext.ResourceKey.create("appendPosition");
```

Comment:
The append position resources sounds like something that is dependent on each specific transaction. If for whatever reason there are multiple transactions present in a ProcessingContext (which is possible when loading and storing events in more than one (bounded) context), then these resources would conflict.

Reply:
Fair point, we need to make sure such a clash doesn't happen. The main reason I had the `ResourceKey` here is for clear sharing with users of the `EventStoreTransaction`.

However, with our recent discussion around the `Context` changes (that made their way in [this](https://github.com/AxonFramework/AxonFramework/pull/3139) PR concerning the `MessageStream.Entry`), I gather we shouldn't share the key, but share the operation to be able to retrieve the resource instead.

Hence, that makes it so that we can move this key into the `DefaultEventStoreTransaction`, **and** provide a method to retrieve the append position from the `EventStoreTransaction`.
What do you think?